// react-say/lib/Say.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Composer = require('./Composer');

var _Composer2 = _interopRequireDefault(_Composer);

var _SayPrimitive = require('./SayPrimitive');

var _SayPrimitive2 = _interopRequireDefault(_SayPrimitive);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Say = function Say(props) {
  return _react2.default.createElement(
    _Composer2.default,
    {
      speechSynthesis: props.speechSynthesis,
      speechSynthesisUtterance: props.speechSynthesisUtterance
    },
    _react2.default.createElement(
      _SayPrimitive2.default,
      {
        lang: props.lang,
        onBoundary: props.onBoundary,
        onEnd: props.onEnd,
        onError: props.onError,
        onStart: props.onStart,
        pitch: props.pitch,
        rate: props.rate,
        speak: props.speak,
        voice: props.voice,
        volume: props.volume
      },
      props.children
    )
  );
};

Say.defaultProps = {
  speechSynthesis: window.speechSynthesis || window.webkitSpeechSynthesis,
  speechSynthesisUtterance: window.SpeechSynthesisUtterance || window.webkitSpeechSynthesisUtterance
};

Say.propTypes = {
  lang: _propTypes2.default.string,
  onBoundary: _propTypes2.default.func,
  onEnd: _propTypes2.default.func,
  onError: _propTypes2.default.func,
  onStart: _propTypes2.default.func,
  pitch: _propTypes2.default.number,
  rate: _propTypes2.default.number,
  speechSynthesis: _propTypes2.default.any,
  speechSynthesisUtterance: _propTypes2.default.any,
  speak: _propTypes2.default.string,
  voice: _propTypes2.default.oneOfType([_propTypes2.default.any, _propTypes2.default.func]),
  volume: _propTypes2.default.number
};

exports.default = Say;
