// react-say/lib/SayButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Composer = require('./Composer');

var _Composer2 = _interopRequireDefault(_Composer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SayButton = function (_React$Component) {
  (0, _inherits3.default)(SayButton, _React$Component);

  function SayButton(props) {
    (0, _classCallCheck3.default)(this, SayButton);

    var _this = (0, _possibleConstructorReturn3.default)(this, (SayButton.__proto__ || (0, _getPrototypeOf2.default)(SayButton)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);

    _this.state = {
      busy: false,
      id: Date.now() + Math.random()
    };
    return _this;
  }

  (0, _createClass3.default)(SayButton, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.props.context.cancel(this.state.id);
    }
  }, {
    key: 'handleClick',
    value: function handleClick() {
      var _this2 = this;

      var props = this.props,
          state = this.state;


      props.onClick && props.onClick(event);

      this.setState(function () {
        return { busy: true };
      }, (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return props.context.speak({
                  id: state.id,
                  lang: props.lang,
                  onBoundary: props.onBoundary,
                  onEnd: props.onEnd,
                  onError: props.onError,
                  onStart: props.onStart,
                  pitch: props.pitch,
                  rate: props.rate,
                  text: props.speak,
                  voice: props.voice,
                  volume: props.volume
                });

              case 2:

                _this2.setState(function () {
                  return {
                    busy: false
                  };
                });

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, _this2);
      })));
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          disabled = _props.disabled,
          busy = this.state.busy;


      return _react2.default.createElement(
        'button',
        {
          disabled: typeof disabled === 'boolean' ? disabled : busy,
          onClick: this.handleClick
        },
        children
      );
    }
  }]);
  return SayButton;
}(_react2.default.Component);

var SayButtonWithContext = function SayButtonWithContext(props) {
  return _react2.default.createElement(
    _Composer2.default,
    {
      speechSynthesis: props.speechSynthesis,
      speechSynthesisUtterance: props.speechSynthesisUtterance
    },
    function (context) {
      return _react2.default.createElement(SayButton, (0, _extends3.default)({ context: context }, props));
    }
  );
};

SayButtonWithContext.propTypes = {
  lang: _propTypes2.default.string,
  onBoundary: _propTypes2.default.func,
  onEnd: _propTypes2.default.func,
  onError: _propTypes2.default.func,
  onStart: _propTypes2.default.func,
  pitch: _propTypes2.default.number,
  rate: _propTypes2.default.number,
  speak: _propTypes2.default.string,
  voice: _propTypes2.default.oneOfType([_propTypes2.default.any, _propTypes2.default.func]),
  volume: _propTypes2.default.number
};

exports.default = SayButtonWithContext;
