// react-say/lib/SayPrimitive.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Context = require('./Context');

var _Context2 = _interopRequireDefault(_Context);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SayPrimitive = function (_React$Component) {
  (0, _inherits3.default)(SayPrimitive, _React$Component);

  function SayPrimitive(props) {
    (0, _classCallCheck3.default)(this, SayPrimitive);

    var _this = (0, _possibleConstructorReturn3.default)(this, (SayPrimitive.__proto__ || (0, _getPrototypeOf2.default)(SayPrimitive)).call(this, props));

    _this.handleBoundary = _this.handleBoundary.bind(_this);
    _this.handleEnd = _this.handleEnd.bind(_this);
    _this.handleError = _this.handleError.bind(_this);
    _this.handleStart = _this.handleStart.bind(_this);

    _this.state = { id: Date.now() + Math.random() };
    return _this;
  }

  (0, _createClass3.default)(SayPrimitive, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      // TODO: Should dequeue self on unmount
      this.unmounted = true;

      this.props.context.cancel(this.state.id).catch(function (err) {
        return 0;
      });
    }
  }, {
    key: 'handleBoundary',
    value: function handleBoundary(event) {
      !this.unmounted && this.props.onBoundary && this.props.onBoundary(event);
    }
  }, {
    key: 'handleEnd',
    value: function handleEnd(event) {
      !this.unmounted && this.props.onEnd && this.props.onEnd(event);
    }
  }, {
    key: 'handleError',
    value: function handleError(event) {
      !this.unmounted && this.props.onError && this.props.onError(event);
    }
  }, {
    key: 'handleStart',
    value: function handleStart(event) {
      !this.unmounted && this.props.onStart && this.props.onStart(event);
    }
  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps) {
      return nextProps.speak !== this.props.speak;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
        var _props, context, lang, pitch, rate, text, voice, volume, id;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _props = _this2.props, context = _props.context, lang = _props.lang, pitch = _props.pitch, rate = _props.rate, text = _props.speak, voice = _props.voice, volume = _props.volume;
                id = _this2.state.id;


                context.speak({
                  id: id,
                  lang: lang,
                  onBoundary: _this2.handleBoundary,
                  onEnd: _this2.handleEnd,
                  onError: _this2.handleError,
                  onStart: _this2.handleStart,
                  pitch: pitch,
                  rate: rate,
                  text: text,
                  voice: voice,
                  volume: volume
                });

              case 3:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, _this2);
      }))().catch(function (err) {
        return console.error(err);
      });

      return false;
    }
  }]);
  return SayPrimitive;
}(_react2.default.Component);

var SayPrimitiveWithContext = function SayPrimitiveWithContext(props) {
  return _react2.default.createElement(
    _Context2.default.Consumer,
    null,
    function (context) {
      return _react2.default.createElement(
        SayPrimitive,
        (0, _extends3.default)({
          context: context
        }, props),
        props.children
      );
    }
  );
};

SayPrimitiveWithContext.propTypes = {
  lang: _propTypes2.default.string,
  pitch: _propTypes2.default.number,
  rate: _propTypes2.default.number,
  onBoundary: _propTypes2.default.func,
  onEnd: _propTypes2.default.func,
  onError: _propTypes2.default.func,
  onStart: _propTypes2.default.func,
  speak: _propTypes2.default.string,
  voice: _propTypes2.default.oneOfType([_propTypes2.default.any, _propTypes2.default.func]),
  volume: _propTypes2.default.number
};

exports.default = SayPrimitiveWithContext;
